### DISCUSSION TITLE
Sensor(s) measuring the opposite of what they should

### DISCUSSION THREAD

**Discussion by neotil1** [srozum/film_camera_tester#12] [Q&A] [answered]
Hi! I've assembled my tester and built sensor 1 and 2! The main menu works flawlessly, but I haven't been able to produce any good readings.

Since sensor 1 didn't work properly, I decided to quickly assemble sensor 2, since it's much simpler and I thought I made some mistake wiring everything up. However, it doesn't work either!

Sensor 2 measures the time it is dark, not the time light is shined into it! When I hold it up to a light source and cover the photodiode for one second, the reading is 1s. So the "normal" state is bright, with the reading being done while the diode is dark...

My Arduino's pinout matches and I've triple checked my wires to sensor 2's board. The photodiode is soldered in the right way (as per the silkscreen on the PCB) and I even tried swapping it around, but that results in a short :D

Is there any simple way I can swap this around? Does anyone know of any more troubleshooting steps? I've already tried two different firmware versions.

TIA!

**Comment by srozum** [comment, marked as answer]
Hi, it's hard to troubleshoot with provided information. Next time try to add photos.

So, if you look at schematics of a sensor, signal pin of a photosensor (SIC553-04) is pulled up by 10K resistor because it has open collector output, which means that default signal is +5V and photosensor pulls it down to GND when triggered. My only guess is that you've got some other type of photosensor.

Check part number and data sheet of your photodiode?

**Reply by neotil1** [reply]
Thank you so much for the quick reply! AliExpress sellers don't really provide datasheets, but after some quick testing with a 5V power supply it's clear I got the wrong photodiodes.

Since I don't have the patience to wait for new parts from China, here's the fix for anyone having the same issue! To invert the photodiode's outputs, all you need is one NPN transistor as well as a 10k resistor:

![20230228_011546](https://user-images.githubusercontent.com/118439425/221719391-df1cb013-bfe8-4756-aaf5-38fce9bc80f5.jpg)

A couple words are in German so here's a legend:
- "hell" = bright (light is shining onto the sensor)
- "dunkel" = dark (no light is shining onto the sensor)
- "gelb" = yellow (signal wires A, B or C)

Here's my hacky way of soldering all this into Sensor 2:
![20230228_011038](https://user-images.githubusercontent.com/118439425/221719775-62112df1-11bb-4ec8-81a2-890b9cd4cb03.jpg)

The 10kOhm resistor is soldered across two of the photodiode's pins. The trace from the yellow wires has to be cut! This is hardly visible in the image. I suggest using the schematic as my soldering may be hard to interpret.

Here's what I did for Sensor 1:
![20230228_005850](https://user-images.githubusercontent.com/118439425/221720127-3c73fd66-bc6b-4b24-aa20-9f1e8e112830.jpg)

Again, traces had to be cut! For this sensor you need to cut all 3 signal traces (A, B, C) and solder 3 resistors and 3 transistors to invert all of the photodiodes. Again, use this image as inspiration only! The right transistor has been FLIPPED!

**Comment by srozum** [comment]
Again, without a photo it's hard to tell. Maybe you soldered it incorrectly.

**Comment by srozum** [comment]
That's cool that you were able come up with a solution. I could make a Firmware version for reversed photodiodes, but not sure if it will work correctly.

If you have wrong photodiodes with unknown hysteresis and sensitivity, plus added transistor key, the timings will be different and Tester might produce wrong results at higher shutter speeds. I'll recommend to find correct ones.

Did you ordered just photosensors or whole modules like shown here https://github.com/srozum/film_camera_tester/discussions/2

Because those modules have same schematics as sensor 1 or 2.

**Reply by neotil1** [reply]
Yeah, I'm going to try and find the right ones, although I don't know how to discern the right ones from the wrong ones. Do you mind sharing where/from which seller you bought yours?

I got mine here: https://a.aliexpress.com/_Ewj7XCv, the board looks identical, but upon closer inspection the description does tell me they are reversed. Gotta pay more attention when ordering next time.

The speeds I've tested so far seem decent, but I'll definitely do some more tomorrow. The transistor I used should be able to switch fast enough so I'm only worried about the potentially incorrect photodiodes... My X-700 constantly produced underexposed images (sometimes capping) at higher shutter speeds, so 1/500->1/750 and 1/1000->1/2300 seems plausible. I haven't been able to test any known good (ish) cameras. It's way past bedtime, so it'll have to wait until tomorrow.

In any case, I have a second set of PCBs and will probably purchase a second set of photodiodes as soon as I'm sure I've found the correct ones.

**Reply by polygonfuture** [reply]
Damn these are the exact same ones I purchased too.

**Comment by srozum** [comment]
Interesting. I bought modules for my units from different sources and all were correct. The only difference I see on product photos that photosensor has red mark on the edge. Maybe that's a difference. But if it works for you then great.

**Reply by polygonfuture** [reply]
@srozum  is your source for these still available?  If so can you simplify things by posting where you purchased them from?

EDIT:  Nevermind I see you are releasing a firmware with inverted logic for sensor 1 and 2

**Comment by srozum** [comment]
I've checked various pages, including ones where I ordered mine photosensors, and they now have this note about reversed output. I don't know what exact IC they using now and what their characteristics are.

Unfortunately, I have no time to experiment with those new modules. Least what I can do is to assemble a Firmware variant for reversed outputs of Sensor 1 and 2. If accuracy stays the same then we are lucky.

Thank you for bringing up this issue and providing solution.

**Reply by neotil1** [reply]
Of course! Thank you for all the help.

If there's anything I can test with an oscilloscope, let me know. I'm also going to contact the AliExpress seller to ask for the part number, so we can hopefully determine the characteristics with the datasheet.

Maybe it would be wise to switch to a non-discontinued part to prevent something like this from happening again? I'm willing to spend some time to get you the necessary info to update the firmware, I just need some intial pointers so I can start my research.

**Reply by neotil1** [reply]
The AliExpress seller just answered my messages: The part number is DS-650B

**Reply by srozum** [reply]
I couldn't find any information nor data sheet for DS-650B, do you?

**Comment by srozum** [comment]
I did some research an found that Kodenshi updated their line of photosensors with parts PIC0103SL and PIC0903SL, which are according to datasheets are exactly the same as SIC553-04 but one has normal open-collector output and another has inverted output.

I believe that PIC0103SL is what currently used in those modules. While PIC0903SL is a new version of SIC553-04.

Either should work in Tester using current PCB designs without modifications, however if you got photo IC with inverted output the firmware have to be updated accordingly.

Firmware version for inverted sensors will be released soon.

**Reply by neotil1** [reply]
Thank you so much for doing the research and updating the firmware, I'm not sure why the AliExpress seller sent me that weird part number. 

It would be nice to know the actual part number of the photo IC in those laser detector modules, but I guess we're going to be left speculating for now

**Reply by srozum** [reply]
Similar ICs also produced by other manufacturers (AMS, TI, Vishai, and others), but those are either have different voltages or different spectral responses or different packaging or different timings. Only Kodenshi photosensors match measured characteristics.
